### PR TITLE
feat: decouple context data format from variable parsing

### PR DESCRIPTION
**Pull request by marstamm**
[ fyi @nikku ] as discussed, a branch to decouple variable format from the parsing. I also added a test that verifies parsing is still works correctly with another Context class

Closes https://github.com/nikku/lezer-feel/issues/17

### CODE REVIEW

File: src/tokens.js [hunk 1]
```diff
@@ -676,9 +777,7 @@ class Variables {
       name,
       tokens: [ ...tokens ],
       children: [ ...children ],
-      context: {
-        ...context
-      },
+      context: context,
```
Comment (nikku) [inline src/tokens.js:780]: We can use the `ES6` shorthand here.

File: src/tokens.js [hunk 2]
```diff
@@ -426,14 +426,118 @@ const dateTimeLiterals = {
...
+   */
+  get(key) {
+    const result = this.value[key];
+
+    if (this.isAtomic(result)) {
+      return result;
+    }
+
+    return new BaseContext(this.value[key]);
+
```
Comment (nikku) [inline src/tokens.js:476]:
```suggestion
```

File: src/tokens.js [hunk 3]
```diff
@@ -730,10 +829,10 @@ function wrap(variables, scopeName, code) {
  *
  * @return { ContextTracker<Variables> }
  */
-export function trackVariables(context = {}) {
+export function trackVariables(context = {}, ContextClass = BaseContext) {
```
Comment (nikku) [inline src/tokens.js:832]: Proposal: Rename `ContextClass` to `VariableContext`

File: src/tokens.js [hunk 4]
```diff
@@ -426,14 +426,118 @@ const dateTimeLiterals = {
 
 const dateTimeIdentifiers = Object.keys(dateTimeLiterals);
 
+
+/**
+ * A basic key-value store to hold context values.
+ */
+export class BaseContext {
```
Comment (nikku) [inline src/tokens.js:433]: Proposal: Rename to `VariableContext`

Comment (marstamm) [reply src/tokens.js]: done: [295dd6e](https://github.com/nikku/lezer-feel/pull/18/commits/295dd6eb52536bc3809a714ade56889a56764e96)

File: src/tokens.js (hunk 2 above)
Comment (marstamm) [reply src/tokens.js]: done: [295dd6e](https://github.com/nikku/lezer-feel/pull/18/commits/295dd6eb52536bc3809a714ade56889a56764e96)

**Comment by nikku** [comment]
Cf. https://github.com/nikku/lezer-feel/pull/18/commits/821a7ce56fc864cfe66a98aee63db37ba579c6b6 for a couple of simplifications that I put on top. The API did change slightly, but only slightly.

File: test/test-feel.js [hunk 5]
```diff
@@ -180,26 +163,45 @@ for (const file of fs.readdirSync(caseDir)) {
...
+
+      for (const { name: testName, run } of tests) {
+
+        const {
+          it,
+          name,
+          context
+        } = parseTest(testName, contextTracker);
+
+        it(name, () => run(createParser(context)));
```
Comment (marstamm) [inline test/test-feel.js:182]: To use the custom Context tracker when executing the test, it should be:

```suggestion
        } = parseTest(testName);

        it(name, () => run(createParser(context), contextTracker));
```

Even with the fix, it still uses the default `VariableContext` sometimes. I was not able to find the root-cause just yet. I will look into it further

Comment (nikku) [reply test/test-feel.js]: Can we create a test case for this, as close to your actual usage scenario?

Comment (nikku) [reply test/test-feel.js]: Fixed via https://github.com/nikku/lezer-feel/pull/18/commits/eafc3e1bccb2545b2bc55bea0dae7fce9f377692.

Comment (marstamm) [reply test/test-feel.js]: > Can we create a test case for this, as close to your actual usage scenario?

Not sure if I have time for it today, but I will look into it

**Comment by marstamm** [comment]
> Even with the fix, it still uses the default VariableContext sometimes. I was not able to find the root-cause just yet. I will look into it further

When no Context was provided, the tests still used the default one. With [48fa4e1](https://github.com/nikku/lezer-feel/pull/18/commits/48fa4e1c5499341b28996b83d46911d4fc9ddfa7), we ensure that the custom context tracker is always used.

We also talked about using the Context class from an instance, without the need of a second parameter. I implemented this with [02fa642](https://github.com/nikku/lezer-feel/pull/18/commits/02fa642acb532077a36d1caa4d6b2b901b11d8a9)

**Comment by nikku** [comment]
Thanks again for updating this contribution! I decided to merge it without the variable context constructor magic (https://github.com/nikku/lezer-feel/pull/18/commits/02fa642acb532077a36d1caa4d6b2b901b11d8a9). Better to be explicit here.

**Comment by nikku** [comment]
This is now released with `v0.17.0`. Thank you for your work on this improvement.

**Pull request merged**